src/context.rs

```diff
@@ -0,0 +1,354 @@
+        let type_id = &TypeId::of::<T>();
+        if !self.plugin_data.contains_key(type_id) {
+            self.add_plugin::<T>();
+        }
+        let data_container = self
+            .plugin_data
+            .get_mut(type_id)
+            .unwrap()
+            .downcast_mut::<T::DataContainer>();
+        match data_container {
+            Some(x) => x,
+            None => panic!("Plugin data container of incorrect type"),
+        }
+    }
+
+    pub fn get_data_container<T: Plugin>(&self) -> Option<&T::DataContainer> {
+        let type_id = &TypeId::of::<T>();
+        if !self.plugin_data.contains_key(type_id) {
+            return None;
+        }
+        let data_container = self
+            .plugin_data
+            .get(type_id)
+            .unwrap()
+            .downcast_ref::<T::DataContainer>();
+        match data_container {
+            Some(x) => Some(x),
+            None => panic!("Plugin data container of incorrect type"),
+        }
+    }
+
+    pub fn get_time(&self) -> f64 {
+        self.time
+    }
+
+    pub fn add_component<T: Component>(&mut self) {
+        T::init(self);
+    }
+
+    fn add_handlers<E: Copy + 'static>(
+        event_handlers: &mut HashMap<TypeId, Box<dyn Any>>,
+        callback: impl Fn(&mut Context, E) + 'static,
+    ) {
+        let callback_vec = event_handlers
+            .entry(TypeId::of::<E>())
+            .or_insert_with(|| Box::<Vec<Rc<EventHandler<E>>>>::default());
+        let callback_vec: &mut Vec<Rc<EventHandler<E>>> = callback_vec.downcast_mut().unwrap();
+        callback_vec.push(Rc::new(callback));
+    }
+
+    pub fn subscribe_to_event<E: Copy + 'static>(
+        &mut self,
+        callback: impl Fn(&mut Context, E) + 'static,
+    ) {
+        Self::add_handlers(&mut self.event_handlers, callback);
+    }
+
+    pub fn subscribe_immediately_to_event<E: Copy + 'static>(
+        &mut self,
+        callback: impl Fn(&mut Context, E) + 'static,
+    ) {
+        Self::add_handlers(&mut self.immediate_event_handlers, callback);
+    }
+
+    fn collect_callbacks<E: Copy + 'static>(
+        event_handlers: &HashMap<TypeId, Box<dyn Any>>,
+        event: E,
+    ) -> Vec<Box<Callback>> {
+        let mut callbacks_to_return = Vec::<Box<Callback>>::new();
+        let callback_vec = event_handlers.get(&TypeId::of::<E>());
+        if let Some(callback_vec) = callback_vec {
+            let callback_vec: &Vec<Rc<EventHandler<E>>> = callback_vec.downcast_ref().unwrap();
+            if !callback_vec.is_empty() {
+                for callback in callback_vec {
+                    let internal_callback = Rc::clone(callback);
+                    callbacks_to_return
+                        .push(Box::new(move |context| internal_callback(context, event)));
+                }
+            }
+        }
+        callbacks_to_return
+    }
+
+    pub fn release_event<E: Copy + 'static>(&mut self, event: E) {
```

Comment:
Is the event emitter stuff new? We may way to split this out

Reply:
Pulled out in latest commit